src/build_context.rs

```diff
@@ -1,133 +1,104 @@
+    //   We also set a None value in definitions.
+    //   Then we create a DefinitionsReferenceValidator that stores the index (value) from this mapping.
+    //   At runtime if it needs to call into the validator it simply looks it up by index.
+    // - Any schema with a `'ref'`: before recursing we store the ref here and set a None
+    //   value in definitions.
+    //   As we recurse if a definitions-reference to this
+    //   Then after recursing and completing the CombinedValidator we replace that None value with the
+    //   newly built validator. Now anything that
+    refs: AHashMap<String, usize>,
+
+    // Here we hold an *optional* reference to a definition validator
+    // Any time we encounter a schema with a `'ref'` we build a DefinitionsReferenceValidator that points
+    // to an index in this Vector
+    // At the end of building we check if any values are None, which indicates an unsatisfied reference
+    definitions: Vec<Option<T>>,
 }
 
 impl<T: Clone + std::fmt::Debug> BuildContext<T> {
-    pub fn new(schema: &PyAny) -> PyResult<Self> {
-        let mut used_refs = AHashSet::new();
-        extract_used_refs(schema, &mut used_refs)?;
+    pub fn new(_schema: &PyAny) -> PyResult<Self> {
         Ok(Self {
-            used_refs,
-            slots: Vec::new(),
-            reusable: AHashMap::new(),
+            refs: AHashMap::new(),
+            definitions: Vec::new(),
         })
     }
 
-    pub fn for_self_schema() -> Self {
-        let mut used_refs = AHashSet::with_capacity(3);
-        // NOTE: we don't call `extract_used_refs` for performance reasons, if more recursive references
-        // are used, they would need to be manually added here.
-        // we use `2` as count to avoid `find_slot` pulling the validator out of slots and returning it directly
-        used_refs.insert("root-schema".to_string());
-        used_refs.insert("ser-schema".to_string());
-        used_refs.insert("inc-ex-type".to_string());
-        Self {
-            used_refs,
-            slots: Vec::new(),
-            reusable: AHashMap::new(),
-        }
-    }
-
-    /// Check whether a ref is already in `reusable` or `slots`, we shouldn't allow repeated refs
-    pub fn ref_already_used(&self, ref_: &str) -> bool {
-        self.reusable.contains_key(ref_) || self.slots.iter().any(|slot| slot.slot_ref == ref_)
-    }
-
-    /// check if a ref is used elsewhere in the schema
-    pub fn ref_used(&self, ref_: &str) -> bool {
-        self.used_refs.contains(ref_)
-    }
-
-    /// check if a ref is used within a given schema
-    pub fn ref_used_within(&self, schema_dict: &PyAny, ref_: &str) -> PyResult<bool> {
-        check_ref_used(schema_dict, ref_)
-    }
-
-    /// add a validator/serializer to `reusable` so it can be cloned and used again elsewhere
-    pub fn store_reusable(&mut self, ref_: String, val_ser: T) {
-        self.reusable.insert(ref_, val_ser);
+    fn get_ref_from_id(&self, ref_id: usize) -> &String {
```

Comment:
this is complicated enough that I think it's worth adding a docstring (`///` comment) to each function.

Reply:
added some docstrings, also. renamed / simplified the funcs